### PR TITLE
Search screen/search screen and pagination

### PR DESCRIPTION
**Pull request by ofekcMoveo**
In this branch I created the search screens - one screen displays the 10 latest searches (saved on the device using userDefualts), and one screen displays the articles we got as a result of the search from the API
 I also implemented pagination in displaying the articles in the tables

### CODE REVIEW

File: Dispatcher/Supporting Files/Utils.swift [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+//  Dispatcher
+//
+//  Created by Ofek Cohen on 03/05/2023.
+//
+
+import UIKit
+
+struct Utils {
+    
+    static func showErrorAlert(_ msg: String) -> UIAlertController {
```
Comment (noamkuMoDev) [inline Dispatcher/Supporting Files/Utils.swift]: The function only returns an UIAlertController and doesn't actually display it on the screen so the func name should be accordingly

File: Dispatcher/Repositories/ArticlesRepository.swift [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    
+    private func buildLatestArticlesRequest(_ pageNumber: Int) -> Request {
+        return Request(
+            baseUrl: APIConstants.latestHeadlinesNewscatcherURL,
+            headers: ["x-api-key": APIConstants.APIkey],
+            parameters: ["countries" : "IL", "page_size": APIConstants.ARTICLES_TO_FETCH_AMOUNT, "page" : pageNumber],
+            method: .get)
+    }
+    
+    func getArticlesByKeywords(_ searchKeywords: String, pageNumber: Int, completionHandler: @escaping ( _ articles: [Article], _ totalPages: Int, _ errorMsg: String?) -> Void) {
```
Comment (noamkuMoDev) [inline Dispatcher/Repositories/ArticlesRepository.swift]: I think a better name will be something like "getArticlesByUserSearchWords" as its a bit more accurate

File: Dispatcher/Repositories/ArticlesRepository.swift [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            case .failure(let error):
+                completionHandler([], 0, error.localizedDescription)
+            }
+        }
+    }
+}
+    
+private func buildSearchArticlesRequest(_ searchKeywords: String, _ pageNumber: Int) -> Request {
+    return Request(
+        baseUrl: APIConstants.searchArticlesNewscatcherURL,
```
Comment (noamkuMoDev) [inline Dispatcher/Repositories/ArticlesRepository.swift]: You have 2 methods in this file that build a Request with mostly similar contents (url, headers, method). Better have a dedicated function that does this for you and accepts properties that can be different (e.g. parameters array) only if there are indeed different parameters to pass into it

File: Dispatcher/Constants/APIConstants.swift [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+//  Constants.swift
+//  Dispatcher
+//
+//  Created by Ofek Cohen on 30/04/2023.
+//
+
+import Foundation
+
+struct APIConstants {
+    static let latestHeadlinesNewscatcherURL = "https://api.newscatcherapi.com/v2/latest_headlines"
```
Comment (noamkuMoDev) [inline Dispatcher/Constants/APIConstants.swift]: Basically the baseUrl is the same (https://api.newscatcherapi.com/v2/) and only the endpoint changes. So better have a baseUrl variable and then have different endpoints variables that you stick onto it when needed while you construct the url string when you build the request

File: Dispatcher/Constants/APIConstants.swift [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    static let searchArticlesNewscatcherURL = "https://api.newscatcherapi.com/v2/search"
+    static let APIkey = "pxIuyKmcbGplzZfNRXbSQOKltrJ7Zgi4gHsvdWoM43g"
+    static let ARTICLES_TO_FETCH_AMOUNT = 12
```
Comment (noamkuMoDev) [inline Dispatcher/Constants/APIConstants.swift]: Stick with the same naming convention - don't have camelCase but then also THIS_WAY and also THISway

File: Dispatcher/Constants/AppConstants.swift [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    
+    //MARK: Search
+    static let latestSearchesCellIdentifier = "latestSearchCell"
+    static let latestSearchCellNibName = "LatestSearchCell"
+    static let LATEST_SEARCHES_AMOUNT = 10
+    static let latestSearchesDefualtsKey = "LatestSearches"
+    
+    
+    //MARK: segues
+    static let fromLatestSearchToResults = "fromLatestSearchToResults"
```
Comment (noamkuMoDev) [inline Dispatcher/Constants/AppConstants.swift:30]: Let's give the segues a dedicated struct of its own, preferably in a separate file as we could potentially have many of them

File: Dispatcher/Constants/AppConstants.swift [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+//
+//  Created by Ofek Cohen on 30/04/2023.
+//
+
+import Foundation
+
+struct AppConstants {
+    //MARK: Global instanses
+    static let articlesRepository = ArticlesRepository()
+    static let defaults = UserDefaults.standard
```
Comment (noamkuMoDev) [inline Dispatcher/Constants/AppConstants.swift]: We want to have a dedicated UserDefaultsManager in which we will hold the let defaults = UserDefaults.standard

File: Dispatcher/UI/Search/SearchScreenViewController.swift [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+//  Created by Ofek Cohen on 02/05/2023.
+//
+
+import UIKit
+
+class SearchScreenViewController: UIViewController {
+
+    @IBOutlet weak var searchTableView: UITableView!
+    @IBOutlet weak var clearButton: UIButton!
+    @IBOutlet weak var babkButton: UIButton!
```
Comment (noamkuMoDev) [inline Dispatcher/UI/Search/SearchScreenViewController.swift]: "backButton" :)

File: Dispatcher/UI/Search/SearchScreenViewController.swift [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+        searchTableView.delegate = self
+        searchTableView.register(UINib(nibName: AppConstants.latestSearchCellNibName, bundle: nil), forCellReuseIdentifier: AppConstants.latestSearchesCellIdentifier)
+    }
+    
+    override func viewWillAppear(_ animated: Bool) {
+        searchScreenViewModel.fetchLatestSearchs()
+        searchTableView.reloadData()
+    }
+    
+    private func customSearchBar() {
```
Comment (noamkuMoDev) [inline Dispatcher/UI/Search/SearchScreenViewController.swift]: not an informative function name - we can't know what does the function do unless we read its code

File: Dispatcher/UI/Search/SearchScreenViewController.swift [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+        customSearchBar()
+       
+        searchTableView.dataSource = self
+        searchTableView.delegate = self
+        searchTableView.register(UINib(nibName: AppConstants.latestSearchCellNibName, bundle: nil), forCellReuseIdentifier: AppConstants.latestSearchesCellIdentifier)
+    }
+    
+    override func viewWillAppear(_ animated: Bool) {
+        searchScreenViewModel.fetchLatestSearchs()
+        searchTableView.reloadData()
```
Comment (noamkuMoDev) [inline Dispatcher/UI/Search/SearchScreenViewController.swift]: We want to reload the tableView only after we are sure that the data was fetched as this can fail

File: Dispatcher/Views/SearchTableCells/LatestSearchCell.swift [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    @IBOutlet weak var searchWordsLabel: UILabel!
+    @IBOutlet weak var removeButton: UIButton!
+    
+    var delegate: LatestSearchCellDelegate?
+    
+    override func awakeFromNib() {
+        super.awakeFromNib()
+        
+        self.contentView.backgroundColor = UIColor(named: "screenBackgroungColor")
+        removeButton.isHidden = true
```
Comment (noamkuMoDev) [inline Dispatcher/Views/SearchTableCells/LatestSearchCell.swift]: Why hide the x button and then when creating the cell then you do hide = false?

File: Dispatcher/UI/Homepage/HomepageViewController.swift [hunk 12]
```diff
@@ -6,14 +6,18 @@
...
     
     @IBOutlet weak var searchButton: UIButton!
     @IBOutlet weak var notificationsButton: UIButton!
     @IBOutlet weak var articlesTableView: UITableView!
+    @IBOutlet var homepageView: UIView!
     
-    let articlesToDisplay = Articles().articles
+    let homepageViewModel = HomepageViewModel()
+    let activityIndicator = UIActivityIndicatorView()
+    var currentPage = 1
```
Comment (noamkuMoDev) [inline Dispatcher/UI/Homepage/HomepageViewController.swift]: Should be saved inside the viewModel and not the HomepageVC

File: Dispatcher/UI/Homepage/HomepageViewController.swift [hunk 13]
```diff
@@ -23,12 +27,35 @@ class HomeViewController: UIViewController {
...
+    private func fetchArticles() {
+        self.activityIndicator.startAnimating()
+        homepageViewModel.getTopArticlesFromAPI(pageNumber: currentPage, completionHandler: { errorMsg in
+            if(errorMsg != nil) {
+                self.present(Utils.showErrorAlert(errorMsg!),animated: true, completion: nil)
+                self.activityIndicator.stopAnimating()
+            } else {
+                DispatchQueue.main.async {
+                    self.activityIndicator.stopAnimating()
+                    self.currentPage += 1
```
Comment (noamkuMoDev) [inline Dispatcher/UI/Homepage/HomepageViewController.swift]: Should happen inside the viewModel once we know we fetched successfully and added the data to our array

File: Dispatcher/UI/Search/SearchScreenViewController.swift [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return searchScreenViewModel.latestSearches.count
+        
+    }
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let currentSearchIndex = (searchScreenViewModel.latestSearches.count - indexPath.row) - 1
+        let search = searchScreenViewModel.latestSearches[currentSearchIndex]
+        
+        if let cell = (tableView.dequeueReusableCell(withIdentifier: AppConstants.latestSearchesCellIdentifier, for: indexPath) as? LatestSearchCell) {
```
Comment (noamkuMoDev) [inline Dispatcher/UI/Search/SearchScreenViewController.swift]: If we can't register the custom cell that we want we should return a regular UITableViewCell() (like you did in the homepage)

File: Dispatcher/UI/Search/SearchScreenViewController.swift [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+            searchBar.text = selectedCell.searchWordsLabel.text
+            performSegue(withIdentifier: AppConstants.fromLatestSearchToResults, sender: self)
+        }
+    }
+}
+
+extension SearchScreenViewController: UITableViewDelegate, UISearchBarDelegate, LatestSearchCellDelegate {
+    func searchCellSelected(_ search: String) {
+        searchBar.text = search
+        performSegue(withIdentifier: AppConstants.fromLatestSearchToResults, sender: self)
```
Comment (noamkuMoDev) [inline Dispatcher/UI/Search/SearchScreenViewController.swift]: We need to display the search results inside the table of the homepage, not navigate to a dedicated screen for showing the results as this creates a confusing UX flow:
user sees general articles -> user searches by words -> user sees wanted results -> user must go back to search screen even though he doesn't want to search anything new -> user sees again general articles

So you need to delete the SearchResultsScreen and display the newly fetched filtered results in the homepage

File: Dispatcher/UI/Search/SearchScreenViewController.swift [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+    }
+    
+    func removeButtonPressed(_ searchToRemove: String) {
```
Comment (noamkuMoDev) [inline Dispatcher/UI/Search/SearchScreenViewController.swift:98]: We need a better name because when implementing the func here it's not clean enough what it's related to. Something like "removeSearchCellBtnPressed"

File: Dispatcher/UI/Search/SearchScreenViewController.swift [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+        searchTextField.backgroundColor = .white
+
+    }
+    
+    
+    @IBAction func backButtonPressed(_ sender: UIButton) {
+        self.dismiss(animated: true)
+    }
+    
+    @IBAction func clearButtonPressed(_ sender: UIButton) {
```
Comment (noamkuMoDev) [inline Dispatcher/UI/Search/SearchScreenViewController.swift]: in the context of the design of the screen it's not clear whether it's the x button of the searchbar or the square clear button of the saved searches.
So a better name will be "clearAllRecentSearchesBtnPressed" or something like that

File: Dispatcher/UI/Search/SearchViewModel.swift [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+//  Dispatcher
+//
+//  Created by Ofek Cohen on 02/05/2023.
+//
+
+import Foundation
+
+class SearchViewModel {
+    var articlesToDisplay: [Article] = []
+    var latestSearches: [String] = []
```
Comment (noamkuMoDev) [inline Dispatcher/UI/Search/SearchViewModel.swift]: Better create a dedicated model for the search object as in the future we might want to add more stuff to it.
So instead of [String] it should be [RecentSearchModel] or so.

File: Dispatcher/UI/Search/SearchViewModel.swift [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+    func articlesAmount() -> Int {
+        return articlesToDisplay.count
+    }
+    
+    func currentArticle(index: Int) -> Article {
+        return articlesToDisplay[index]
+    }
+    
+    func fetchLatestSearchs() {
+        latestSearches = (AppConstants.defaults.array(forKey: AppConstants.latestSearchesDefualtsKey) as? [String]) ?? []
```
Comment (noamkuMoDev) [inline Dispatcher/UI/Search/SearchViewModel.swift]: We want a search repository that will communicate with our UserDefaultsManager to comply with the MVVM+repository architecture that we have in our project

File: Dispatcher/UI/Homepage/HomepageViewModel.swift [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+class HomepageViewModel {
+    
+    var articlesToDisplay: [Article] = []
+    var totalResultsPages = 1
+    var isPaginating = false
+     
+    func getTopArticlesFromAPI(pageNumber: Int, completionHandler: @escaping (_ errorMsg: String?) -> Void) {
+        if(isPaginating == false) {
+            isPaginating = true
+            AppConstants.articlesRepository.getLatestArticles(pageNumber: pageNumber) { articles, totalPages,  errorMsg in
```
Comment (abenartz) [inline Dispatcher/UI/Homepage/HomepageViewModel.swift]: ArticleRepository (a class) should not be inside AppConstants(file global constant values).
You have two options:
1. Create an instance of the repository inside your ViewModel.
2. Make ArticleRepository singleton.

What is better for our case here?

File: Dispatcher/UI/Homepage/HomepageViewModel.swift [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+import UIKit
+
+class HomepageViewModel {
+    
+    var articlesToDisplay: [Article] = []
+    var totalResultsPages = 1
+    var isPaginating = false
+     
+    func getTopArticlesFromAPI(pageNumber: Int, completionHandler: @escaping (_ errorMsg: String?) -> Void) {
+        if(isPaginating == false) {
```
Comment (abenartz) [inline Dispatcher/UI/Homepage/HomepageViewModel.swift]: Check here also if the paging was exhausted (when you won't get more results)

**Review by abenartz — CHANGES_REQUESTED** [review]

File: Dispatcher/Supporting Files/Utils.swift [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+               return image
+            }
+        }
+    }
+
+    return UIImage()
+}
+
+
+func createErrorAlert(_ msg: String) -> UIAlertController {
```
Comment (noamkuMoDev) [inline Dispatcher/Supporting Files/Utils.swift:31]: * The naming convention would be to use "getErrorAlert" rather than "createErrorAlert"
* It would be nicer to read when calling this function if the signature was: getErrorAlert(withMessage msg: String)

File: Dispatcher/Supporting Files/AppDelegate.swift [hunk 23]
```diff
@@ -14,6 +14,13 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
 
 
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
+        let response = readApiKeyFromConfigFile()
```
Comment (noamkuMoDev) [inline Dispatcher/Supporting Files/AppDelegate.swift]: Instead of "response" a better naming would be what it actually is - apiKeyValue

File: Dispatcher/Supporting Files/Config.json [hunk 24]
```diff
@@ -0,0 +1,3 @@
+{
+    "api_key": "oE0GNhaEFfOy9eAygqEaq4P9ir8hqxrIui7u20H5_e4"
```
Comment (noamkuMoDev) [inline Dispatcher/Supporting Files/Config.json:2]: Need to add the Config.json to the .gitignore file so that it wouldn't be uploaded to github

File: Dispatcher/Repositories/ArticlesRepository.swift [hunk 25]
```diff
@@ -0,0 +1,89 @@
...
+               
+                completionHandler(articles, Int(dataResult.totalPages), nil)
+            case .failure(let error):
+                completionHandler([], 0, error.localizedDescription)
+            }
+        }
+    }
+    
+    private func buildRequest(_ searchKeywords: String? , _ pageNumber: Int) -> Request {
+        if(searchKeywords != nil) {
```
Comment (noamkuMoDev) [inline Dispatcher/Repositories/ArticlesRepository.swift]: When you need to use the variable that you are checking inside the { } it's better to use "if let" instead of "!= nil" because then you don't need to force-unwrap the variable using "!".
Even though it won't fail because you indeed checked it's not null, we still don't like to use the "!"

File: Dispatcher/Constants/AppConstants.swift [hunk 26]
```diff
@@ -0,0 +1,40 @@
...
+//  Created by Ofek Cohen on 30/04/2023.
+//
+
+import Foundation
+
+struct AppConstants {
+    
+    //MARK: Homepage articles table
+    static let tableRowHight = 40
+    static let topHeadlinesHeaderText = "Top Headlines in Israel"
```
Comment (noamkuMoDev) [inline Dispatcher/Constants/AppConstants.swift]: Better have a separate file for text constants

File: Dispatcher/Constants/AppConstants.swift [hunk 27]
```diff
@@ -0,0 +1,40 @@
...
+    
+    //MARK: Search
+    static let latestSearchesAmount = 10
+    static let latestSearchesDefualtsKey = "LatestSearches"
```
Comment (noamkuMoDev) [inline Dispatcher/Constants/AppConstants.swift]: Better have a separate struct for UserDefaultsKeys

File: Dispatcher/Networking/AlamofireManager.swift [hunk 28]
```diff
@@ -0,0 +1,29 @@
...
+//  Created by Ofek Cohen on 01/05/2023.
+//
+
+import Foundation
+import Alamofire
+
+class AlamofireManager {
+    
+    func sendRequest<T:Decodable>(_ request: Request, completionHandler: @escaping (Result<T, Error>) -> Void) {
+        if let encoded = request.urlWithParams.addingPercentEncoding(withAllowedCharacters: .urlFragmentAllowed),let url = URL(string: encoded) {
```
Comment (noamkuMoDev) [inline Dispatcher/Networking/AlamofireManager.swift:14]: What happens in case the "if let" fails?

File: Dispatcher/Supporting Files/StringErrorExtension.swift [hunk 29]
```diff
@@ -0,0 +1,12 @@
+//
+//  StringError.swift
+//  Dispatcher
+//
+//  Created by Ofek Cohen on 09/05/2023.
+//
+
+import Foundation
+
+extension String: LocalizedError {
```
Comment (noamkuMoDev) [inline Dispatcher/Supporting Files/StringErrorExtension.swift]: Why did you need to add this?

File: Dispatcher/UI/Homepage/HomepageViewModel.swift [hunk 30]
```diff
@@ -0,0 +1,50 @@
...
+    
+    var articlesToDisplay: [Article] = []
+    var totalResultsPages = 1
+    var currentPage = 1
+    var isPaginating = false
+    
+    private init () {}
+     
+    func getArticlesToDisplay(completionHandler: @escaping (_ errorMsg: String?) -> Void) {
+        if(isPaginating == false && currentPage <= totalResultsPages) {
```
Comment (noamkuMoDev) [inline Dispatcher/UI/Homepage/HomepageViewModel.swift]: Better convention is to check for false by: !isPagination

File: Dispatcher/UI/Homepage/HomepageViewModel.swift [hunk 31]
```diff
@@ -0,0 +1,50 @@
...
+            isPaginating = true
+            ArticlesRepository.shared.getArticlesFromApi(pageNumber: currentPage) { articles, totalPages,  errorMsg in
+                if (errorMsg != nil) {
+                    completionHandler(errorMsg!)
```
Comment (noamkuMoDev) [inline Dispatcher/UI/Homepage/HomepageViewModel.swift]: If we failed then we never reset the isPaginating back to false?

File: Dispatcher/Repositories/ArticlesRepository.swift [hunk 32]
```diff
@@ -0,0 +1,89 @@
...
+
+class ArticlesRepository {
+    
+    static let shared = ArticlesRepository()
+    let alamofireManager = AlamofireManager()
+    
+    private init() {}
+
+    func getArticlesFromApi(pageNumber: Int, completionHandler: @escaping (_ articles: [Article], _ totalPages: Int, _ errorMsg: String?) -> Void) {
+        let request = buildRequest(nil, pageNumber)
```
Comment (noamkuMoDev) [inline Dispatcher/Repositories/ArticlesRepository.swift]: When you don't provide names for the parameters that you send in it's much less readable and harder to understand. You actually have to go into the buildRequest function itself to understand that in this particular case you don't have search words and that's why you send nil into it.

File: Dispatcher/Repositories/SearchRepository.swift [hunk 33]
```diff
@@ -0,0 +1,34 @@
...
+//  Dispatcher
+//
+//  Created by Ofek Cohen on 09/05/2023.
+//
+
+import Foundation
+
+class SearchRepository {
+    
+    static let shared = SearchRepository()
```
Comment (noamkuMoDev) [inline Dispatcher/Repositories/SearchRepository.swift:12]: Why did you create it as a singleton?

File: Dispatcher/Persistence/UserDefaultsManager.swift [hunk 34]
```diff
@@ -0,0 +1,24 @@
+//
+//  UserDefaultsManager.swift
```
Comment (noamkuMoDev) [inline Dispatcher/Persistence/UserDefaultsManager.swift:2]: Since this is a Manager file, we would want the functions inside it to be as general as possible (like in the AlamofireManager) so we can use it wherever we need throughout the app.
We don't want to have to create a dedicated function for every data that we will need to save to the user defaults (e.g. if the user saw onboarding or not - will we create a "saveSeenOnboarding" function as well?)

File: Dispatcher/UI/Homepage/HomepageViewController.swift [hunk 35]
```diff
@@ -0,0 +1,145 @@
...
+        return homepageViewModel.articlesAmount()
+        
+    }
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        
+        let currentArticle = homepageViewModel.getArticleByIndex(index: indexPath.row)
+        
+        if let cell = (tableView.dequeueReusableCell(withIdentifier: TableCellsIdentifiers.articleCellIdentifier, for: indexPath) as? ArticleCell) {
+            DispatchQueue.main.async {
```
Comment (noamkuMoDev) [inline Dispatcher/UI/Homepage/HomepageViewController.swift]: The cellForRowAt is a synchronic function that is always executed on the main thread automatically

File: Dispatcher/UI/Homepage/HomepageViewController.swift [hunk 36]
```diff
@@ -0,0 +1,145 @@
...
+            return cell
+        } else {
+            return UITableViewCell()
+        }
+    }
+    
+    func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
+        if indexPath.row == (homepageViewModel.articlesToDisplay.count - 2) {
+            if (homepageViewModel.currentPage < homepageViewModel.totalResultsPages) {
+                getArticles()
```
Comment (noamkuMoDev) [inline Dispatcher/UI/Homepage/HomepageViewController.swift:107]: Eventually we reload the entire tableView just because we added a few more articles (possibly even just 1) to our existing data array (could already have even 100 items).
This is wasteful in resources.
Instead need to use insertRows to just add the new rows to the tableview and leave the existing rows untouched

File: Dispatcher/Repositories/ArticlesRepository.swift [hunk 37]
```diff
@@ -0,0 +1,89 @@
...
+                    }
+                }
+                completionHandler(articles, dataResult.totalPages, nil)
+            case .failure(let error):
+                completionHandler([], 0, error.localizedDescription)
+            }
+        }
+    }
+    
+    func getArticlesByUserSearchWords(_ searchKeywords: String, pageNumber: Int, completionHandler: @escaping ( _ articles: [Article], _ totalPages: Int, _ errorMsg: String?) -> Void) {
```
Comment (noamkuMoDev) [inline Dispatcher/Repositories/ArticlesRepository.swift]: getArticlesByUserSearchWords & getArticlesFromApi are pretty much exactly the same aside from the searchKeywords. So no need for 2 functions

File: Dispatcher/Constants/APIConstants.swift [hunk 38]
```diff
@@ -0,0 +1,15 @@
...
+//  Dispatcher
+//
+//  Created by Ofek Cohen on 30/04/2023.
+//
+
+import Foundation
+
+struct APIConstants {
+    static var apiKey: String = ""
+    static let latestHeadlinesNewscatcherURL = "https://api.newscatcherapi.com/v2"
```
Comment (noamkuMoDev) [inline Dispatcher/Constants/APIConstants.swift]: This is the same url for both variables.
Call it "baseUrl" instead.
And save the different endpoints

File: Dispatcher/Repositories/SearchRepository.swift (hunk 33 above)
Comment (noamkuMoDev) [inline Dispatcher/Repositories/SearchRepository.swift]: Why is it a singleton? Only the SeachViewModel uses it

Comment (ofekcMoveo) [reply Dispatcher/Repositories/SearchRepository.swift]: I did it a singleton because the ArticlesRepository is a singleton and I thought that it should be the same in order to be consistent

Comment (noamkuMoDev) [reply Dispatcher/Repositories/SearchRepository.swift]: הסינגלטון הוא פיתרון שימושי למצבים שבהם אנחנו חייבים שהקלאס יהיה זמין ונגיש לנו ברחבי האפליקציה והדאטה שהוא מחזיק יהיה זהה מכל מקום שניגש אליו - זה עונה על קונספט של single source of truth
אבל חיסרון מרכזי של הסינגלטון הוא שברגע שהוא נוצר לראשונה - הוא יחייה לנצח כל עוד האפליקציה בחיים, גם אם כבר אין לנו בו צורך או שימוש, והוא ממשיך לגזול משאבים של זיכרון שעולה לנו בפרפורמנס של האפליקציה.
ולכן אנחנו לא ניצור סתם סינגלטונים, אלא רק אם יש באמת צורך בכך. במקרה פה למשל, רק ה-SearchViewModel צריך להשתמש ב-SearchRepository ולכן אין לנו צורך ליצור ממנו סינגלטון

**Review by noamkuMoDev — APPROVED** [review]

**Review by abenartz — APPROVED** [review]
